src/Stratis.Bitcoin.Features.Dns/UnreliablePeerBehavior.cs

```diff
@@ -0,0 +1,136 @@
+
+        /// <summary>Information about node's chain.</summary>
+        private readonly IChainState chainState;
+
+        /// <summary>Logger factory to create loggers.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Handle the lifetime of a peer.</summary>
+        private readonly IPeerBanning peerBanning;
+
+        /// <summary>The node settings.</summary>
+        private readonly NodeSettings nodeSettings;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public UnreliablePeerBehavior(Network network, IChainState chainState, ILoggerFactory loggerFactory, IPeerBanning peerBanning, NodeSettings nodeSettings)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(chainState, nameof(chainState));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(peerBanning, nameof(nodeSettings));
+            Guard.NotNull(nodeSettings, nameof(nodeSettings));
+
+            this.network = network;
+            this.chainState = chainState;
+            this.loggerFactory = loggerFactory;
+            this.peerBanning = peerBanning;
+            this.nodeSettings = nodeSettings;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public override object Clone()
+        {
+            return new UnreliablePeerBehavior(this.network, this.chainState, this.loggerFactory, this.peerBanning, this.nodeSettings);
+        }
+
+        /// <inheritdoc />
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync);
+        }
+
+        /// <inheritdoc />
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceivedAsync);
+        }
+
+        /// <summary>
+        /// Event handler that is called when the node receives a network message from the attached peer.
+        /// </summary>
+        /// <param name="peer">Peer that sent us the message.</param>
+        /// <param name="message">Received message.</param>
+        /// <remarks>
+        /// This handler only cares about "version" messages, which are only sent once per node.
+        /// </remarks>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            if (message.Message.Payload is VersionPayload version)
+            {
+                // If current node is on POS, and ProvenHeaders is activated, check if current connected peer can serve Proven Headers.
+                // If it can't, disconnect from him and ban for few minutes
+                if (IsProvenHeaderActivated())
+                {
+                    if (!CanServeProvenHeader(version))
```

Comment:
No need for nesting, condition can be done in 1 line

Reply:
true, i refactored the if conditions into external methods and forget to simplify the blocks